pallet/staking/src/lib.rs

```diff
@@ -946,6 +823,57 @@ where
 	}
 }
 
+/// Staking rate limiter.
+#[derive(Clone, Debug, PartialEq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub enum RateLimiter {
+	/// Positive balance.
+	Pos(Balance),
+	/// Negative balance.
+	Neg(Balance),
+}
+impl RateLimiter {
+	fn flow_in(self, amount: Balance, limit: Balance) -> Option<Self> {
+		match self {
+			Self::Pos(v) => v.checked_add(amount).filter(|&v| v <= limit).map(Self::Pos),
+			Self::Neg(v) =>
+				if v >= amount {
+					Some(Self::Neg(v - amount))
+				} else {
+					let v = amount - v;
+
+					if v <= limit {
+						Some(Self::Pos(v))
+					} else {
+						None
```

Comment:
Do we need to update self here, as there are side effects under other conditions?